### PR TITLE
ocaml-cstruct: 1.6.0 -> 2.3.0

### PR DESCRIPTION
**Pull request by sternenseemann**
cc @vbgl 

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [x] Tested compilation of all pkgs that depend on this change
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

This commit also refactors the expression.

### CODE REVIEW

File: pkgs/development/ocaml-modules/cstruct/default.nix [hunk 1]
```diff
@@ -1,20 +1,40 @@
-{stdenv, writeText, fetchurl, ocaml, ocplib-endian, sexplib_p4, findlib,
- async_p4 ? null, lwt ? null, camlp4}:
+{stdenv, buildOcaml, fetchFromGitHub, writeText,
+ ocaml, ocplib-endian, sexplib_p4, findlib, ounit, camlp4,
```
Comment (hannesm) [inline pkgs/development/ocaml-modules/cstruct/default.nix]: to jump in here, cstruct 2.3.0 has dropped camlp4 support -- thus this is no longer a dependency.

from the opam file (https://github.com/ocaml/opam-repository/blob/master/packages/cstruct/cstruct.2.3.0/opam):
```
depends: [
  "ocamlfind" {build}
  "ounit"     {test}
  "ocplib-endian"
  "sexplib"
  "base-bytes"
]
depopts: [
  "ppx_tools"
  "async"
  "lwt"
]
```

not sure what your `sexplib_p4` is (I suspect a sexplib with camlp4 support), but cstruct does no longer need this, a sexplib without p4 support is fine!

Comment (sternenseemann) [inline pkgs/development/ocaml-modules/cstruct/default.nix]: Thank you! Implementing your suggestions seems to fix the build problem with enabled async support I encountered.

**Comment by vbgl** [comment]
This breaks `ocamlPackages.cstruct`: can we avoid that?

**Comment by sternenseemann** [comment]
@vbgl no, I did some testing, it does not work.

This is the error with standard `ocamlPackages`:

```
configuring
ocamlfind: Package `ppx_tools.metaquot' not found
W: Field 'pkg_ppx_tools_metaquot' is not set: Command ''/nix/store/pxrr9pa01bazbcjy4kdf6kyvr8il70cd-ocaml-findlib-1.6.1/bin/ocamlfind' query -format %d ppx_tools.metaquot > '/tmp/nix-build-ocaml-cstruct-2.3.0.drv-0/oasis-5b264d.txt'' terminated with error code 2
E: Cannot find findlib package ppx_tools.metaquot
E: Failure("1 configuration error")
```

Even with `withPpx = false` this error does not occur in `ocamlPackages_4_02`.

**Comment by sternenseemann** [comment]
@vbgl Do you think, I'd be better to put the new version into a separate package? I guess this is necessary, I wouldn't mind it.

**Comment by vbgl** [comment]
Done in 2d136204c47985c9518105e1af9e14810c8319ec

**Comment by sternenseemann** [comment]
@vbgl So this can be closed?

**Pull request closed without merging**